### PR TITLE
cas login and access log

### PR DESCRIPTION
**Pull request by DoraALin**
1. cas login config.
user who access nsqadmin api need cas login, api access via Basic auth; 
2. login user access log.
acces log before admin action notify 
Signed-off-by: 元守 <lulin@youzan.com>

### CODE REVIEW

**Comment by DoraALin** [comment]
@absolute8511 user auth interface applied, for your review

File: nsqadmin/nsqadmin.go [hunk 1]
```diff
@@ -36,6 +37,52 @@ func New(opts *Options) *NSQAdmin {
 		notifications: make(chan *AdminAction),
 	}
 
+	if opts.CASUrl != "" {
```
Comment (absolute8511) [inline nsqadmin/nsqadmin.go]: This should be abstracted into auth module

File: nsqadmin/http.go [hunk 2]
```diff
@@ -1098,6 +1215,209 @@ func (c TopicsByHourlyPubsize) Less(i, j int) bool {
 	return l > r
 }
 
+type CasResponse struct {
```
Comment (absolute8511) [inline nsqadmin/http.go]: Should move all auth related code to a new file.

File: nsqadmin/http.go [hunk 3]
```diff
@@ -23,8 +24,17 @@ import (
 	"github.com/youzan/nsq/internal/http_api"
 	"github.com/youzan/nsq/internal/protocol"
 	"github.com/youzan/nsq/internal/version"
+	"github.com/gorilla/sessions"
+	"reflect"
+	"errors"
 )
 
+var store = sessions.NewCookieStore([]byte("cas-authen"))
```
Comment (absolute8511) [inline nsqadmin/http.go]: Maybe sessions should go into the auth model.

File: nsqadmin/http.go [hunk 4]
```diff
@@ -100,22 +110,110 @@ func NewHTTPServer(ctx *Context) *httpServer {
...
 	router.Handle("GET", "/api/statistics", http_api.Decorate(s.statisticsHandler, log, http_api.V1))
 	router.Handle("GET", "/api/statistics/:sortBy", http_api.Decorate(s.statisticsHandler, log, http_api.V1))
 	router.Handle("GET", "/api/cluster/stats", http_api.Decorate(s.clusterStatsHandler, log, http_api.V1))
-
+	router.Handle("GET", "/api/oauth/cas/callback", http_api.Decorate(s.casAuthCallbackHandler, log, http_api.V1))
+	router.Handle("GET", "/api/oauth/cas/callback/logout", http_api.Decorate(s.casAuthCallbackLogoutHandler, log, http_api.V1))
 	return s
 }
 
+func (s *httpServer) getExistingUserInfo(req *http.Request) (IUserAuth, error) {
```
Comment (absolute8511) [inline nsqadmin/http.go]: Move this to the auth model.

**Comment by absolute8511** [comment]
We also hide the button for topic Delete, channel Empty, Delete if not login.

File: nsqadmin/http.go [hunk 5]
```diff
@@ -100,22 +110,110 @@ func NewHTTPServer(ctx *Context) *httpServer {
...
 	router.Handle("GET", "/api/statistics/:sortBy", http_api.Decorate(s.statisticsHandler, log, http_api.V1))
 	router.Handle("GET", "/api/cluster/stats", http_api.Decorate(s.clusterStatsHandler, log, http_api.V1))
-
+	router.Handle("GET", "/api/oauth/cas/callback", http_api.Decorate(s.casAuthCallbackHandler, log, http_api.V1))
+	router.Handle("GET", "/api/oauth/cas/callback/logout:", http_api.Decorate(s.casAuthCallbackLogoutHandler, log, http_api.V1))
 	return s
 }
 
+func (s *httpServer) getExistingUserInfo(req *http.Request) (IUserAuth, error) {
+	session, err := store.Get(req, "session-userInfo")
```
Comment (absolute8511) [inline nsqadmin/http.go]: Maybe move this to auth module.

**Pull request merged**